stdlib/public/core/RangeReplaceableCollection.swift

```diff
@@ -1139,4 +1139,37 @@ extension RangeReplaceableCollection {
     // self = try filter { try !predicate($0) }
     self = try Self(self.lazy.filter { try !shouldBeRemoved($0) })
   }
+
+  /// Removes and returns all the elements that satisfy the given predicate.
+  ///
+  /// Use this method in place of `removeAll(where:)` to retain the removed
+  /// elements for further processing.
+  ///
+  ///     var phrase = "The rain in Spain stays mainly in the plain."
+  ///
+  ///     let vowels: Set<Character> = ["a", "e", "i", "o", "u"]
+  ///     let extracted = phrase.removeAll(where: { vowels.contains($0) })
+  ///     // phrase == "Th rn n Spn stys mnly n th pln."
+  ///     // extracted == "eaiiaiaaioeai"
+  ///
+  /// - Parameter shouldBeExtracted: A closure that takes an element of the
+  ///   sequence as its argument and returns a Boolean value indicating
+  ///   whether the element should be extracted from the collection.
+  ///
+  /// - Complexity: O(*n*), where *n* is the length of the collection.
+  @inlinable
+  public mutating func extractAll(
+    where shouldBeExtracted: (Element) throws -> Bool
+  ) rethrows -> Self {
+    var extracted = Self()
+    extracted.reserveCapacity(self.count)
```

Comment:
Is this too much overkill? Another option is to place all the extracted element into an array and then return `Self(array)` at the end. Not sure what the efficiency implications are there.